File: go-server/services/subject_service.go
package services

import (
	"context"
	"errors"
	"fmt"
	"go-server/models"
	"go-server/repositories"
	"go-server/utils"

	"github.com/sirupsen/logrus"
	"gorm.io/gorm"
)

type SubjectService interface {
	GenerateSubjects(courseID uint) ([]models.Subject, error)
	GetSubjectsByCourseID(courseID uint) ([]models.Subject, error)
	GetSubjectByID(subjectID uint) (*models.Subject, error)
	CreateSubjects(subjects []models.Subject) error
	UpdateSubject(subjectID uint, name string) (*models.Subject, error)
	DeleteSubject(subjectID uint) error
	DeleteSubjectsByCourseID(courseID uint) error
}

type SubjectServiceImpl struct {
	subjectRepo  repositories.SubjectRepository
	courseRepo   repositories.CourseRepository
	geminiHelper *utils.GeminiHelper
}

func NewSubjectService(subjectRepo repositories.SubjectRepository, courseRepo repositories.CourseRepository) SubjectService {
	return &SubjectServiceImpl{
		subjectRepo:  subjectRepo,
		courseRepo:   courseRepo,
		geminiHelper: utils.NewGeminiHelper(),
	}
}

// GenerateSubjects generates subjects for a course using Gemini AI
func (s *SubjectServiceImpl) GenerateSubjects(courseID uint) ([]models.Subject, error) {
	// Get course details
	course := &models.Course{}
	if err := s.courseRepo.GetByID(courseID, course); err != nil {
		if err == gorm.ErrRecordNotFound {
			return nil, errors.New("course not found")
		}
		return nil, errors.New("failed to get course")
	}

	// Check if subjects already exist
	existingSubjects, err := s.subjectRepo.GetSubjectsByCourseID(courseID)
	if err != nil {
		return nil, errors.New("failed to check existing subjects")
	}

	// If updating, clear existing subjects first
	if course.HasSubjects && len(existingSubjects) > 0 {
		if err := s.subjectRepo.DeleteSubjectsByCourseID(courseID); err != nil {
			logrus.Errorf("Error deleting existing subjects: %v", err)
		}
	}

	// Generate subjects using Gemini
	ctx := context.Background()
	prompt := fmt.Sprintf(`Based on the course '%s' with description '%s', 
	generate a list of relevant subjects that should be included in this course.
	Consider the following:
	1. If it's a school/college/university course, align with their typical curriculum
	2. Don't include the course name as a subject
	3. Keep subjects relevant and practical
	4. Generate maximum 5 core subjects for the course`, course.Name, course.Description)

	subjectNames, err := s.geminiHelper.GenerateSubjects(ctx, course.Name, course.Description)
	if err != nil {
		logrus.Errorf("Error generating subjects with Gemini: %v", err)
		return nil, errors.New("failed to generate subjects")
	}

	// Create subject models
	var subjects []models.Subject
	for _, name := range subjectNames {
		subjects = append(subjects, models.Subject{
			Name:     name,
			CourseID: courseID,
		})
	}

	// Save subjects
	if err := s.subjectRepo.CreateSubjects(subjects); err != nil {
		logrus.Errorf("Error saving subjects: %v", err)
		return nil, errors.New("failed to save subjects")
	}

	// Update course has_subjects flag
	if err := s.courseRepo.UpdateHasSubjects(courseID, true); err != nil {
		logrus.Errorf("Error updating has_subjects flag: %v", err)
		// Don't fail the operation for this
	}

	logrus.Infof("Generated %d subjects for course: %s", len(subjects), course.Name)
	return subjects, nil
}

// GetSubjectsByCourseID returns all subjects for a course
func (s *SubjectServiceImpl) GetSubjectsByCourseID(courseID uint) ([]models.Subject, error) {
	subjects, err := s.subjectRepo.GetSubjectsByCourseID(courseID)
	if err != nil {
		logrus.Errorf("Error getting subjects by course ID: %v", err)
		return nil, errors.New("failed to get subjects")
	}
	return subjects, nil
}

// GetSubjectByID returns a specific subject
func (s *SubjectServiceImpl) GetSubjectByID(subjectID uint) (*models.Subject, error) {
	subject := &models.Subject{}
	if err := s.subjectRepo.GetByID(subjectID, subject); err != nil {
		if err == gorm.ErrRecordNotFound {
			return nil, errors.New("subject not found")
		}
		logrus.Errorf("Error getting subject: %v", err)
		return nil, errors.New("failed to get subject")
	}
	return subject, nil
}

// CreateSubjects creates multiple subjects
func (s *SubjectServiceImpl) CreateSubjects(subjects []models.Subject) error {
	if err := s.subjectRepo.CreateSubjects(subjects); err != nil {
		logrus.Errorf("Error creating subjects: %v", err)
		return errors.New("failed to create subjects")
	}
	return nil
}

// UpdateSubject updates a subject's name
func (s *SubjectServiceImpl) UpdateSubject(subjectID uint, name string) (*models.Subject, error) {
	subject := &models.Subject{}
	if err := s.subjectRepo.GetByID(subjectID, subject); err != nil {
		if err == gorm.ErrRecordNotFound {
			return nil, errors.New("subject not found")
		}
		return nil, errors.New("failed to get subject")
	}

	subject.Name = utils.SanitizeString(name)
	if err := s.subjectRepo.Update(subject); err != nil {
		logrus.Errorf("Error updating subject: %v", err)
		return nil, errors.New("failed to update subject")
	}

	return subject, nil
}

// DeleteSubject deletes a subject
func (s *SubjectServiceImpl) DeleteSubject(subjectID uint) error {
	if err := s.subjectRepo.Delete(subjectID, &models.Subject{}); err != nil {
		logrus.Errorf("Error deleting subject: %v", err)
		return errors.New("failed to delete subject")
	}
	return nil
}

// DeleteSubjectsByCourseID deletes all subjects for a course
func (s *SubjectServiceImpl) DeleteSubjectsByCourseID(courseID uint) error {
	if err := s.subjectRepo.DeleteSubjectsByCourseID(courseID); err != nil {
		logrus.Errorf("Error deleting subjects by course ID: %v", err)
		return errors.New("failed to delete subjects")
	}
	return nil
}

File: python-server/services/subject_service.py
# services/subject_service.py
from repositories.subject_repo import SubjectRepository
from models.subject import Subject
from models.schemas import SubjectContent
from repositories.course_repo import CourseRepository
from utils.gemini_helper import GeminiHelper, extract_sql_query
from utils.gemini_image_generation_helper import GeminiImageGenerator
from utils.unified_storage_helper import storage_helper
from sqlalchemy.orm import Session
import logging

logger = logging.getLogger(__name__)

class SubjectService:
    def __init__(self, db: Session):
        self.subject_repo = SubjectRepository(db)
        self.course_repo = CourseRepository(db)
        self.storage_helper = storage_helper

    def generate_subjects(self, course_id):
        course = self.course_repo.get_course_by_id(course_id)
        gemini_helper = GeminiHelper()

        if not course:
            raise Exception("Course not found")

        prompt = f"""Based on the course '{course.name}' with description '{course.description}', 
        generate a list of  relevant subjects that should be included in this course.
        Consider the following:
        1. If it's a school/college/university course, align with their typical curriculum
        2. Don't include the course name as a subject
        3. Keep subjects relevant and practical
        4. Generate maximum 5 core subjects for the course
        """
        
        try:
            response = gemini_helper.generate_content(
                prompt,
                response_schema=SubjectContent
            )
            
            # If updating, clear existing subjects first
            if course.has_subjects:
                self.subject_repo.delete_subjects_by_course_id(course_id)
            
            # Initialize image generator
            image_generator = GeminiImageGenerator()
            created_subjects = []
            
            for subject_name in response['subjects']:
                subject = Subject(
                    course_id=course_id,
                    name=subject_name
                )
                created_subject = self.subject_repo.add_subject(subject)
                created_subjects.append(created_subject)
                
                # Try to generate an image for each subject
                try:
                    # Generate image
                    image_bytes = image_generator.generate_subject_image(subject_name, course.name)
                    
                    if image_bytes:
                        # Upload to Azure Storage
                        image_path = f"courses/{course_id}/subjects/{created_subject.id}/cover"
                        image_url = self.storage_helper.upload_image(image_bytes, image_path)
                        
                        # Update subject with image URL
                        self.subject_repo.update_subject_image(created_subject.id, image_url)
                except Exception as img_error:
                    # Log but don't fail if image generation fails
                    logger.error(f"Error generating subject image: {str(img_error)}")
            
            # Mark that the course has subjects
            self.course_repo.set_has_subjects(course_id, True)

            return {"message": "Subjects generated successfully"}
            
        except Exception as e:
            raise Exception(f"Error generating subjects: {e}")

    def get_subjects_by_course_id(self, course_id):
        subjects = self.subject_repo.get_subjects_by_course_id(course_id)
        return [subject.to_dict() for subject in subjects]

    def create_subject(self, course_id, name):
        logger.info(f"Creating new subject for course_id: {course_id}")
        
        # Verify course exists
        course = self.course_repo.get_course_by_id(course_id)
        if not course:
            logger.error(f"Course not found for id: {course_id}")
            raise Exception("Course not found")
            
        try:
            subject = self.subject_repo.create_subject(course_id, name)
            
            # If this is first subject, mark course as having subjects
            if not course.has_subjects:
                self.course_repo.set_has_subjects(course_id, True)
                
            # Try to generate an image using environment API key
            try:
                # Initialize image generator
                image_generator = GeminiImageGenerator()
                
                # Generate image
                image_bytes = image_generator.generate_subject_image(name, course.name)
                
                if image_bytes:
                    # Upload to Azure Storage
                    image_path = f"courses/{course_id}/subjects/{subject.id}/cover"
                    image_url = self.storage_helper.upload_image(image_bytes, image_path)
                    
                    # Update subject with image URL
                    self.subject_repo.update_subject_image(subject.id, image_url)
            except Exception as img_error:
                # Log but don't fail if image generation fails
                logger.error(f"Error generating subject image: {str(img_error)}")
                
            return subject.to_dict()
        except Exception as e:
            logger.error(f"Error creating subject: {str(e)}")
            raise Exception(f"Error creating subject: {str(e)}")

    def update_subject(self, subject_id, name):
        logger.info(f"Updating subject id: {subject_id}")
        
        try:
            subject = self.subject_repo.update_subject(subject_id, name)
            if subject:
                return subject.to_dict()
            else:
                logger.error(f"Subject not found for id: {subject_id}")
                raise Exception("Subject not found")
        except Exception as e:
            logger.error(f"Error updating subject: {str(e)}")
            raise Exception(f"Error updating subject: {str(e)}")
            
    def delete_subject(self, subject_id):
        logger.info(f"Deleting subject id: {subject_id}")
        
        try:
            success = self.subject_repo.delete_subject(subject_id)
            if success:
                return {"message": "Subject deleted successfully"}
            else:
                logger.error(f"Subject not found for id: {subject_id}")
                raise Exception("Subject not found")
        except Exception as e:
            logger.error(f"Error deleting subject: {str(e)}")
            raise Exception(f"Error deleting subject: {str(e)}")